Refactor codecs

sdeleuze opened the pull request:
This commit introduces the following changes:
- MessageToByteEncoder/Decoder renamed to Encoder/Decoder
- JsonObjectEncoder/Decoder are now used directly in
JacksonJsonEncoder/Decoder
- Codec uses MimeType instead of MediaType since they
are not specific to HTTP
- Default MimeType are now managed thanks to
Encoder/Decoder#getSupportedMimeTypes()
- AbstractEncoder/Decoder takes care of generic MimeType related behavior

poutsma commented on src/main/java/org/springframework/reactive/codec/Codec.java:
```diff
@@ -0,0 +1,57 @@
+			Object... hints);
+
+	/**
+	 * Return the list of {@link MediaType} objects supported by this decoder.
```
MimeType instead of MediaType

sdeleuze replied:
Fixed thanks

sdeleuze commented:
I have updated this PR as discussed during the meeting. We keep the encoder/decoder split and the fact that they encode/decode from/to `ByteBuffer`.

I have just modified the name since `MessageToByteEncoder/MessageToByteDecoder` is not always accurate (a pre/postProcessor can keep the same data granularity) and `Encoder/Decoder` is more simple, short and still accurate in our context since we have decided to always deal with `ByteBuffer`.

Can I merge it to master, or do you have any other feedbacks?

sdeleuze commented:
Last update to use `Bytes` interface instead of `ByteBuffer` in order to let us experiment with various byte buffer implementation as discussed in Issue #7.

rstoyanchev commented:
I've processed the 1st commit with the codec refactoring and also improved the 'content-type' and 'accept' header handling including actual content negotiation. 

As for the `Bytes` abstraction let's use a branch to work out something more definitive.

sdeleuze commented:
Thanks for merging that in master, and for the content negotiation refinements.

I will create a branch for the `Bytes` abstraction.

For `MediaType`versus `MimeType`, I have seen the additional conversion steps it requires (cf. your commit https://github.com/spring-projects/spring-reactive/commit/cb0ce46b11c044578378054f4d1f6e5bf0b8a06d), sorry for the pain, but I am wondering if keeping that as it is currently in master couldn't be nice to highlight the changes that may be needed in Spring Framework 5 + it would be nice to be able to use codec in our upcoming messaging support without using http packages ...

Speaking about Spring Framework 5, I have several questions in mind:
 - Should we provide some facilities to create a `MediaType` from a `MimeType`?
 - Should we move constant like `APPLICATION_JSON` and `APPLICATION_JSON_VALUE` from `MediaType` to `MimeType` (`MediaType` extends `MimeType` so I guess that would be a non breaking change, and since these values have nothing related with quality parameters), and deprecate those constants in `MimeTypeUtils` in 4.3 in order to remove them in 5.0 in favor of the `MimeType` ones?

rstoyanchev commented:
bq. I am wondering if keeping that as it is currently in master couldn't be nice to highlight the changes that may be needed

I don't think nice is how I thought of it :) ... all joke aside though it is a valid design question. I just don't think open design questions should go into master. It's much easier to think and experiment with solutions in a branch that everyone can look at and agree on vs being faced with technical debt while working on unrelated changes.
